src/dbnode/integration/query_cancellation_client_test.go

```diff
@@ -0,0 +1,233 @@
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package integration
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/m3db/m3/src/dbnode/client"
+	"github.com/m3db/m3/src/dbnode/namespace"
+	"github.com/m3db/m3/src/dbnode/retention"
+	"github.com/m3db/m3/src/dbnode/storage/index"
+	"github.com/m3db/m3/src/m3ninx/idx"
+	"github.com/m3db/m3/src/x/ident"
+	xsync "github.com/m3db/m3/src/x/sync"
+	xtime "github.com/m3db/m3/src/x/time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/zap"
+)
+
+func TestQueryCancellationAndDeadlinesClient(t *testing.T) {
+	if testing.Short() {
+		t.SkipNow() // Just skip if we're doing a short run
+	}
+
+	var (
+		retentionPeriod = 2 * time.Hour
+		dataBlockSize   = time.Hour
+		indexBlockSize  = 2 * time.Hour
+		bufferFuture    = 20 * time.Minute
+		bufferPast      = 10 * time.Minute
+	)
+
+	// Test setup
+	md, err := namespace.NewMetadata(testNamespaces[0],
+		namespace.NewOptions().
+			SetRetentionOptions(
+				retention.NewOptions().
+					SetRetentionPeriod(retentionPeriod).
+					SetBufferPast(bufferPast).
+					SetBufferFuture(bufferFuture).
+					SetBlockSize(dataBlockSize)).
+			SetIndexOptions(
+				namespace.NewIndexOptions().
+					SetBlockSize(indexBlockSize).SetEnabled(true)).
+			SetColdWritesEnabled(true))
+	require.NoError(t, err)
+
+	var (
+		hostQueueWorkerPools     []*mockWorkerPool
+		hostQueueWorkerPoolsLock sync.Mutex
+	)
+
+	testOpts := NewTestOptions(t).
+		SetNamespaces([]namespace.Metadata{md}).
+		SetWriteNewSeriesAsync(true).
+		SetCustomClientAdminOptions([]client.CustomAdminOption{
+			client.CustomAdminOption(func(opts client.AdminOptions) client.AdminOptions {
+				return opts.SetHostQueueNewPooledWorkerFn(func(
+					opts xsync.NewPooledWorkerOptions,
+				) (xsync.PooledWorkerPool, error) {
+					workerPoolOpts := xsync.NewPooledWorkerPoolOptions().
+						SetGrowOnDemand(true).
+						SetKillWorkerProbability(0.01).
```

Comment:
hmm I'm not seeing how this is used in this test

Reply:
Yeah I can remove this, it was copied. The mock worker pool needs a real backing implementation.